File: modern/_1.18.2/src/main/java/cn/enaium/noexpensive/callback/AnvilTakeOutputCallback.java
package cn.enaium.noexpensive.callback;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.item.ItemStack;

/**
 * @author Enaium
 */
public interface AnvilTakeOutputCallback {
    Event<AnvilTakeOutputCallback> EVENT = EventFactory.createArrayBacked(AnvilTakeOutputCallback.class, (listeners) -> (output) -> {
        for (AnvilTakeOutputCallback listener : listeners) {
            listener.interact(output);
        }
    });

    void interact(ItemStack output);
}

File: modern/_1.21.3/src/main/kotlin/cn/enaium/noexpensive/callback/AnvilTakeOutputCallback.kt
package cn.enaium.noexpensive.callback

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack

/**
 * @author Enaium
 */
fun interface AnvilTakeOutputCallback {
    fun interact(output: ItemStack, player: PlayerEntity)

    companion object {
        val EVENT: Event<AnvilTakeOutputCallback> =
            EventFactory.createArrayBacked(AnvilTakeOutputCallback::class.java) { listeners: Array<AnvilTakeOutputCallback> ->
                AnvilTakeOutputCallback { output: ItemStack, player: PlayerEntity ->
                    for (listener in listeners) {
                        listener.interact(output, player)
                    }
                }
            }
    }
}
